rflx/pyrflx/typevalue.py

```diff
@@ -461,119 +465,138 @@ def accepted_type(self) -> type:
-        """
-        :param value: Bitstring representation of the message
-        :param length: use length parameter if complete
-        length of the message is less than 1 Byte
-        """
+    def assign_bitvalue(self, value: Bitstring, offset: int = 0) -> None:
 
         current_field_name = self._next_field(model.INITIAL.name)
-        field_first_in_bitstr = 0
-        field_length = 0
+        last_field_first_in_bitstr = current_field_first_in_bitstr = 0 + offset
+        current_field_length = 0
 
-        while current_field_name != model.FINAL.name and (
-            field_first_in_bitstr + field_length
-        ) <= len(str(value)):
+        def get_current_pos_in_bitstr(field_name: str) -> int:
 
-            current_field = self._fields[current_field_name]
-            if isinstance(current_field.typeval, OpaqueValue) and not self._has_length(
-                current_field_name
-            ):
+            # if the node is a virtual node i.e. has the same first as the previous node
+            # set the current pos in bitstring back to the first position of its predecessor
+            this_first = self._fields[field_name].first
+            prev_first = self._fields[self._prev_field(field_name)].first
+            assert isinstance(prev_first, Number)
+            assert isinstance(this_first, Number)
+
+            return (
+                last_field_first_in_bitstr
+                if prev_first.value == this_first.value
+                else current_field_first_in_bitstr
+            )
+
+        def set_field_without_length(field_name: str, field: MessageValue.Field) -> Tuple[int, int]:
+
+            last_pos_in_bitstr = current_pos_in_bitstring = current_field_first_in_bitstr
+            assert isinstance(field.typeval, OpaqueValue)
+            field.first = self._get_first(field_name)
+            self.set(field_name, value[current_pos_in_bitstring:])
+            field.length = Number(field.typeval.size)
+
+            return last_pos_in_bitstr, current_pos_in_bitstring
+
+        def set_field_with_length(field_name: str, field_length: int) -> Tuple[int, int]:
+
+            last_pos_in_bitstr = current_pos_in_bitstring = get_current_pos_in_bitstr(field_name)
+
+            if field_length <= 8:
+
+                self.set(
+                    field_name,
+                    value[current_pos_in_bitstring : current_pos_in_bitstring + field_length],
+                )
+                current_pos_in_bitstring += field_length
 
-                self._fields[current_field_name].first = self._get_first(current_field_name)
-                self.set(current_field_name, value[field_first_in_bitstr:])
-                self._fields[current_field_name].length = Number(current_field.typeval.length)
             else:
-                assert isinstance(current_field.length, Number)
-                field_length = current_field.length.value
+                bytes_used_for_field = field_length // 8 + 1
```

Comment:
It works as intended since python does not raise an error if the second index is less than the first index an a slice and returns an empty string. I will change it for better readability.

Reply:
But this also changes the range in the following for loop adding more bits to `field_bits`. If this is intended maybe the name `bytes_used_for_field` is misleading then?

Reply:
I added an explicit handling for fields which have a length of a multiple of 8.

Reply:
Fixed by 4fb9eac